llvm/tools/llvm-xargs/llvm-xargs.cpp

```diff
@@ -0,0 +1,140 @@
+//===- llvm-xargs.cpp - Command lines execution ---------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This source is utility to execute command lines. The specified command will
+// be invoked as many times as necessary to use up the list of input items.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/FileSystem.h"
+#include "llvm/Support/LineIterator.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/Program.h"
+#include "llvm/Support/SystemUtils.h"
+
+using namespace llvm;
+
+static cl::opt<std::string> InputFileList{
+    cl::Positional, cl::Required, cl::desc("<input list of files>"),
+    cl::init("-"), cl::value_desc("filename")};
+
+static cl::opt<std::string> InputCommand{
+    cl::Positional, cl::Required, cl::desc("<command>"), cl::init("echo"),
+    cl::value_desc("command")};
+
+static cl::opt<std::string> Replace{
+    "in-replace",
+    cl::desc("Replace R in command with names read from standard input; if R "
+             "is unspecified, assume {}"),
+    cl::init("{}"), cl::value_desc("R")};
+
+static cl::opt<std::string> OutReplace{
+    "out-replace",
+    cl::desc("Replace R in command with temporary filename for writing outputs "
+             "from executed command"),
+    cl::init(""), cl::value_desc("R")};
+
+// Emit list of produced files for better integration with other tools.
+static cl::opt<std::string> OutputFileList{
+    "out-file-list", cl::desc("Specify filename for list of outputs."),
+    cl::value_desc("filename"), cl::init("")};
+
+static void error(const Twine &Msg) {
+  errs() << "llvm-xargs: " << Msg << '\n';
+  exit(1);
+}
+
+static void error(std::error_code EC, const Twine &Prefix) {
+  if (EC)
+    error(Prefix + ": " + EC.message());
+}
+
+int main(int argc, char **argv) {
+
+  cl::ParseCommandLineOptions(
+      argc, argv,
+      "llvm-xargs: Execute specified command as many times as\n"
+      "necessary to use up the list of input items.\n"
+      "Usage:\n"
+      "llvm-xargs a.list 'echo {}'\n"
+      "NOTE: filenames containing blanks or newlines and commands containig\n"
+      "redirects are not supported by llvm-xargs yet.\n");
+
+  ExitOnError ExitOnErr("llvm-xargs: ");
+  std::unique_ptr<MemoryBuffer> MB =
+      ExitOnErr(errorOrToExpected(MemoryBuffer::getFileOrSTDIN(InputFileList)));
+  std::string FileList = "";
+
+  // Split input command into program and args.
+  std::string CurCommand = InputCommand;
+  StringRef ResCommand = CurCommand;
+  SmallVector<StringRef, 4> Args;
+  ResCommand.split(Args, /*Separator*/ " ", /*MaxSplit*/ -1,
```

Comment:
This doesn't support executables with spaces, right? Quotes and escape sequences can make it a bit complicated.
This can actually be avoided if llvm-xargs (like *nix xargs) could take a command as a set of positional arguments, instead of a single string argument.

Reply:
Yes, executables and filenames with spaces aren't supported. I left this info in tool's help.
An idea with set of positional arguments sounds good, but I'm not fully understand how it will work with filenames/executables with spaces, because positional arguments still separated by spaces, right? Not sure that tool can figure out which of input positional arguments is something with spaces and which is just two arguments.
Please correct me if I'm missing something :)

Reply:
> Yes, executables and filenames with spaces aren't supported. I left this info in tool's help.
> An idea with set of positional arguments sounds good, but I'm not fully understand how it will work with filenames/executables with spaces, because positional arguments still separated by spaces, right? Not sure that tool can figure out which of input positional arguments is something with spaces and which is just two arguments.
> Please correct me if I'm missing something :)

When a program is invoked, it's a caller's responsibility to "split" positional arguments.
For example, `"/home/I look for troubles/bin/llvm-xargs" foo "bar baz"` is splitted by a caller (shell, or clang driver in this case), and llvm-xargs tool sees its arguments as:
```
argv[0] = "/home/I look for troubles/bin/llvm-xargs"
argv[1] = "foo"
argv[2] = "bar baz"
```

Specifically, instead of:
```llvm-xargs --out-replace=%t.out %t.list --out-file-list=%t.out.list "copy {} %t.out"```
you can write:
```llvm-xargs --out-replace=%t.out --not-sure-what-this-is=%t.list --out-file-list=%t.out.list -- copy {} %t.out```
Here, "copy", "{}", and "%t.out" are three distinct positional arguments. "--" is important, as it separates arguments for xargs and arguments for the command.

Reply:
Ah, ok. But this behavior also should be documented, I guess, in help for example. I'll give it a try, thanks!

Reply:
Added.